mars/optimization/logical/core.py

```diff
@@ -130,6 +131,77 @@ def _replace_node(self, original_node: EntityType, new_node: EntityType):
+        graph: Optional[EntityGraph],
+        removed_nodes: Optional[Set[EntityType]],
+        new_results: Optional[List[Entity]] = None,
+    ):
+        """
+        Replace the subgraph from the self._graph represented by a list of nodes with input graph.
+        It will delete the nodes in removed_nodes with all linked edges first, and then add (or update if it's still
+        existed in self._graph) the nodes and edges of the input graph.
+
+        Parameters
+        ----------
+        graph : EntityGraph, optional
+            The input graph. If it's none, no new node and edge will be added.
+        removed_nodes : Set[EntityType], optional
+            The nodes to be removed. All the edges connected with them are removed as well.
+        new_results : List[EntityType], optional, default None
+            The updated results of the graph. If it's None, then the results will not be updated.
+
+        Raises
+        ------
+        ReplaceSubgraphError
+            If the input key of the removed node's successor can't be found in the subgraph.
+            Or some of the nodes of the subgraph are in removed ones.
+        """
+        infected_successors = set()
+
+        output_to_node = dict()
+        removed_nodes = removed_nodes or set()
+        if graph is not None:
+            # Add the output key -> node of the subgraph
+            for node in graph.iter_nodes():
+                if node in removed_nodes:
+                    raise ValueError(f"The node {node} is in the removed set")
+                for output in node.outputs:
+                    output_to_node[output.key] = node
+
+        for node in removed_nodes:
+            for infected_successor in self._graph.iter_successors(node):
+                if infected_successor not in removed_nodes:
+                    infected_successors.add(infected_successor)
+        # Check whether infected successors' inputs are in subgraph
+        for infected_successor in infected_successors:
+            for inp in infected_successor.inputs:
+                if inp.key not in output_to_node:
+                    raise ValueError(
+                        f"The output {inp} of node {infected_successor} is missing in the subgraph"
+                    )
+        for node in removed_nodes:
+            self._graph.remove_node(node)
+
+        if graph is None:
+            return
+
+        # Add the output key -> node of the original graph
+        for node in self._graph.iter_nodes():
+            for output in node.outputs:
+                output_to_node[output.key] = node
+
+        for node in graph.iter_nodes():
+            self._graph.add_node(node)
+
+        for node in itertools.chain(graph.iter_nodes(), infected_successors):
+            for inp in node.inputs:
+                pred_node = output_to_node[inp.key]
+                self._graph.add_edge(pred_node, node)
+
+        if new_results is not None:
+            self._graph.results = new_results.copy()
```

Comment:
Should we merge `new_results` into existing ones instead of accepting it as the whole set? also
```suggestion
            self._graph.results = list(new_results)
```

Reply:
If we use merging strategy, then some additional logic should be implemented. i.e. there are 3 nodes v1 -> v2 -> v3 in the original graph and v3's outputs are in `results`. Now we use v3->v4 to replace the graph (the grpah will be v1 -> v2 -> v3 -> v4 in final). The new results should be v4's outputs. However during the replacement there's no any removing actions, but we should know which result will be removed.  So if using merging instead of replacing, a `deleted_results` should be passed to this function.

Reply:
Fixed. Add a `new_results` parameter.